tripy/tests/frontend/trace/ops/test_storage.py

```diff
@@ -16,62 +16,43 @@
-        assert storage.data.memref_value.address_space == runtime.PointerType.host
-        assert storage.device.kind == "cpu"
 
-    def test_gpu_storage(self):
-        data = Array([1, 2, 3], shape=(3,), dtype=None, device=tp.device("gpu"))
+    @pytest.mark.parametrize("device", ["cpu", "gpu"])
+    def test_from_memref(self, device):
+        module = np if device == "cpu" else cp
+        data = memref.create_memref_view(module.ones((2, 2), dtype=module.float32))
         storage = Storage([], [], data)
-        assert isinstance(storage.data.memref_value, runtime.MemRefValue)
-        assert storage.data.memref_value.address_space == runtime.PointerType.device
-        assert storage.device.kind == "gpu"
+        assert storage.has_memref is True
+        assert storage.dtype == tp.float32
+        assert storage.shape == (2, 2)
+        assert storage.device.kind == device
 
-    @pytest.mark.parametrize("dtype", DATA_TYPES.values())
-    def test_dtype(self, dtype):
-        # Given a int/float data list, store data with requested data type.
-        if dtype not in {tp.float32, tp.int32, tp.int64}:
-            pytest.skip(f"List to tp.Array conversion only supports float32, int32, int64. Got {dtype}")
-        arr = [1.0, 2.0, 3.0] if dtype == tp.float32 else [1, 2, 3]
-        data = Array(arr, shape=(3,), dtype=dtype, device=None)
+    def test_from_list(self):
+        data = [[1.0, 2.0], [3.0, 4.0]]
         storage = Storage([], [], data)
-        assert storage.dtype == dtype
-        assert storage.dtype.name == dtype.name
-        assert storage.dtype.itemsize == dtype.itemsize
+        assert storage.has_memref is False
+        assert storage.dtype == tp.float32
+        assert storage.shape == (2, 2)
+        assert storage.device.kind == "gpu"
 
-    @pytest.mark.parametrize("dtype", DATA_TYPES.values())
-    def test_mlir_conversion(self, dtype):
-        # Given a int/float data list, store data with requested data type.
-        if dtype not in {tp.float32, tp.int32, tp.int64}:
-            pytest.skip(f"List to tp.Array conversion only supports float32, int32, int64. Got {dtype}")
-        arr = [1.0, 2.0, 3.0] if dtype == tp.float32 else [1, 2, 3]
-        data = Array(arr, shape=(3,), dtype=dtype, device=None)
-        storage = Storage([], [TraceTensor("t0", None, dtype, None, None, 1, shape=[3])], data)
-        with mlir_utils.make_ir_context(), ir.Location.unknown():
-            flat_ir = FlatIR()
-            fir_outputs = [out.to_flat_ir() for out in storage.outputs]
-            storage.to_flat_ir([], fir_outputs)
-            flat_ir.integrate_subgraph([], fir_outputs)
-            mlir_outputs = flat_ir.ops[0].to_mlir(operands=[])
-            assert mlir_outputs[0].value.type.element_type == mlir_utils.get_mlir_dtype(dtype)
+    def test_empty_list(self):
+        data = [[]]
+        storage = Storage([], [], data, dtype=tp.float16)
+        assert storage.has_memref is True
```

Comment:
No, it's not because of dtype. It's because MLIR `DenseElementAttr` cannot handle an empty list, thus we have to create an empty memref, see the relevant code in `storage.py` with comment `special case: empty tensor`.